hackathon_site/hardware/serializers.py

```diff
@@ -46,9 +54,128 @@ def get_unique_hardware_count(obj: Category):
+            )
+        )
+        category_counts = dict()
+        error_messages = []
+        for (hardware, requested_quantity) in requested_hardware.items():
+            try:
+                team_hardware = team_unreturned_orders.get(id=hardware.id)
+            except ObjectDoesNotExist:
+                team_hardware = None
+            team_hardware_count = getattr(team_hardware, "past_order_count", 0)
+            if hardware.max_per_team < (team_hardware_count + requested_quantity):
+                error_messages.append("Hardware {} limit reached".format(hardware.name))
+            for category in hardware.categories.all():
+                category_counts[category] = (
+                    category_counts.get(category, 0) + team_hardware_count
+                )
+        for (category, count) in category_counts.items():
+            if category.max_per_team < count:
+                error_messages.append("Category {} limit reached".format(category.name))
+        if error_messages:
+            error_message = "; ".join(error_messages)
+            raise serializers.ValidationError(error_message)
+        return data
+
+    def create(self, validated_data):
+        requested_hardware = self.merge_requests(
+            hardware_requests=validated_data["hardware"]
+        )
+        relevant_past_order_items = (
+            OrderItem.objects.filter(hardware__in=requested_hardware.keys(),)
+            .exclude(order__status="Cancelled")
+            .select_related("hardware")
+        )
+        team_past_hardware_counts = Counter(
+            (
+                map(
+                    lambda item: item.hardware,
+                    relevant_past_order_items.filter(
+                        order__team=validated_data["team_id"],
+                    ).all(),
+                )
+            )
+        )
+        all_past_hardware_counts = Counter(
+            (map(lambda item: item.hardware, relevant_past_order_items.all(),))
+        )
+        new_order = None
+        unfulfilled_hardware_requests = dict()
+        for (hardware, requested_quantity) in requested_hardware.items():
+            allowed_quantity = min(
+                hardware.max_per_team - team_past_hardware_counts[hardware],
+                hardware.quantity_available - all_past_hardware_counts[hardware],
+            )
+            if allowed_quantity <= 0:
+                unfulfilled_hardware_requests[hardware] = requested_quantity
+                continue
+            if new_order is None:
+                new_order = Order.objects.create(
+                    team=validated_data["team_id"], status="Submitted"
+                )
+            if allowed_quantity < requested_quantity:
+                unfulfilled_hardware_requests[hardware] = (
+                    requested_quantity - allowed_quantity
+                )
+            for _ in itertools.repeat(None, min(allowed_quantity, requested_quantity)):
```

Comment:
I'll let most of the other itertools/functools stuff slide, but this is Python.

```python
for _ in range(min(allowed_quantity, requested_quantity)):
    ...
```

Even better, for efficiency:

```python
order_items = [OrderItem(order=new_order, hardware=hardware) for _ in range(min(allowed_quantity, requested_quantity))]
OrderItem.objects.bulk_create(order_items)
```

If you wanted to be even more efficient, keep a list of order items to create outside of this loop, then insert them all in a single database query at the end with `OrderItem.objects.bulk_create()`.

Reply:
On second thought, definitely do that last thing of creating them all with a single bulk create. This will significantly cut down on the number of database queries made when everyone submits orders at once.

Reply:
yup done